From 55d01be9962e8f3c0d0c6016d33431d42647ba98 Mon Sep 17 00:00:00 2001
From: Vadim Pasternak <vadimp@nvidia.com>
Date: Mon, 19 Jun 2023 09:53:41 +0000
Subject: [PATCH backport 6.1.42 37/85] mlxsw: core_hwmon: Align modules label
 name assignment according to the MTCAP sensor counter

For some new devices MTCAP register provides the counter of ASIC
ambient sensors plus additional number of platform sensors.
In such case 'sensor count' will be greater then 1 and modules labels
will be shifted by the number of platform sensors.

Thus utilities sensors will expose incorrect modules labels.
For example, temperatures for module#1, module#2 will be exposed like:
front panel 002:  +37.0C  (crit = +70.0C, emerg = +75.0C)
front panel 003:  +47.0C  (crit = +70.0C, emerg = +75.0C)
instead of:
front panel 002:  +37.0C  (crit = +70.0C, emerg = +75.0C)
front panel 003:  +47.0C  (crit = +70.0C, emerg = +75.0C)

Set 'index' used in label name according to the 'sensor_count' value.

Signed-off-by: Vadim Pasternak <vadimp@nvidia.com>
---
 drivers/net/ethernet/mellanox/mlxsw/core_hwmon.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/net/ethernet/mellanox/mlxsw/core_hwmon.c b/drivers/net/ethernet/mellanox/mlxsw/core_hwmon.c
index 70735068cf29..3455aa749f73 100644
--- a/drivers/net/ethernet/mellanox/mlxsw/core_hwmon.c
+++ b/drivers/net/ethernet/mellanox/mlxsw/core_hwmon.c
@@ -403,6 +403,19 @@ mlxsw_hwmon_module_temp_label_show(struct device *dev,
 {
 	struct mlxsw_hwmon_attr *mlxsw_hwmon_attr =
 			container_of(attr, struct mlxsw_hwmon_attr, dev_attr);
+	struct mlxsw_hwmon_dev *mlxsw_hwmon_dev;
+	int index = mlxsw_hwmon_attr->type_index;
+
+	mlxsw_hwmon_dev = mlxsw_hwmon_attr->mlxsw_hwmon_dev;
+	/* For some devices 'sensor count' provides the number of ASIC sensor
+	 * plus additional platform sensors. Set 'index' used in label name
+	 * according to the 'sensor_count' value to align label name with the
+	 * module index.
+	 */
+	if (mlxsw_hwmon_dev->sensor_count > 1)
+		index += 1 - mlxsw_hwmon_dev->sensor_count;
+	if (strlen(mlxsw_hwmon_dev->name))
+		index += 1;
 
 	return sprintf(buf, "front panel %03u\n",
 		       mlxsw_hwmon_attr->type_index);
-- 
2.20.1

